### ISSUE TITLE
[provision] Do not auto return server  if tests result was failed

### ISSUE THREAD

**Issue by guazhang** [teemtee/tmt#2582]
we should keep the test server (test environment ) a short time to debug.

the function more like `/distribution/reservesys`  param `RESERVE_IF_FAIL`

**Comment by idorax** [comment]
What's the `provision` method you once used? @guazhang

**Comment by guazhang** [comment]
> What's the `provision` method you once used? @guazhang
beaker.
after tmt finished the `finish` step,  The test machine will return whether the result is pass or not. but I hope it can reserved if result is failed.

**Comment by idorax** [comment]
Looks  a parameter (e.g. `reserve-if-fail`) should be added to support the feature. Hence, `tmt` should be able to detect the test results when `provision` is `beaker`. @happz What do you think?

**Comment by happz** [comment]
I recall discussions about splitting `finish` into two steps, to allow `finish` scripts to happen without returning machines, for example. This issue feels related, sort of an optional `provision` cleanup. But I don't recall details or any issues. @psss @lukaszachy might know more.

Yes, some kind of `--reserve-if-fail` option sounds like something to invent, preferably making it standardized behavior across all `provision` plugins. I'm not sure where this option should belong, but it feels like something `tmt run` should take care of rather than individual plugins: plugins expose "acquire" and "release" operations, and something must decide whether "release" needs to be called - based on inspecting results. This spans several steps (`execute` provides results, `provision` owns guests), therefore the entity "seeing" all these steps would be the one to decide, and that's `tmt run`, or `tmt.base.Run` class in my view.

**Comment by psss** [comment]
Yeah, agree with @happz. We need to separate `finish` and `cleanup` of guests. I'll try to kick off something soon. Then the clean up part could be marked as optional under an `if-fail` condition.

@guazhang, by the way, not sure whether you know about the feature: The `login` command can be used to access the guest upon failure in the following way:

    tmt run login --when fail

This can be useful when executing tests directly from user laptop.

**Comment by lukaszachy** [comment]
I'd say that this is not specific to 'beaker' - having the way to keep machine running (skip the 'guest cleanup) is handy for 'virtual'  or 'minute' as well.

**Comment by guazhang** [comment]
> tmt run login --when fail

yes, this method also help to keep server running, but I usual use the `login` function  to check error after one case finished, and save logs if the case failed,  see issue [#2587 ](https://github.com/teemtee/tmt/issues/2587),

```
login  -w fail -w error -s execute -t -c 'bash $TMT_TREE/script/log_check.sh
```

**Comment by lukaszachy** [comment]
Also it is not only about 'fail' of the test. User might want to investigate an error too, also errors in prepare/finish steps (e.g. investigating why prepare script didn't run successfully)

**Comment by idorax** [comment]
Looks a parameter `--reserve` is required when provision is beaker?  @lukaszachy @happz @guazhang

**Comment by lukaszachy** [comment]
Not really. tmt has one dummy task scheduled in the Beaker and runs everything on its own. Beaker itself isn't aware of any other activity.

tmt (mrack provider) could use such a flag, it can be set on an already running job too. 

But tmt doesn't need to. All what is required is to skip the guest cleanup part when conditions are met. 

AFAIK that is required also if the flag is used as guest cleanup cancels the running job. I believe that job cancellation wins over reserve (as beaker  reserve flag usecase is just keeping machines available after all tasks are done)


Try some tmt run with beaker and check the job in Beaker webui. You'll see how it's done.

**Comment by lukaszachy** [comment]
Eh, I've followed my own advice and checked the generated job.xml - mrack already adds `<reservesys duration="86400" when="always"/>` so the 'keep machine on failure' logic has to happen in tmt (Beaker really doesn't know what is being run there)

**Comment by happz** [comment]
Correct, the same applies to Artemis, it has absolutely no idea what's happening to a machine, and a reserved machine is no different than a non-reserved one, it's just a machine the user decided to not destroy, yet. This indeed needs tmt-level handling.

**Comment by psss** [comment]
The `cleanup` step has been implemented in #3904 so we should now be unblocked. I guess the new option should go under the `tmt run` subcommand and should work across all `provision` plugins, not limited to `beaker` only:

    tmt run --reserve-if-fail

Or we could extend the `--skip` option with a condition. It could look like this:

    tmt run --skip cleanup --skip-when fail,error
    tmt run --skip cleanup --skip-when-not pass

Seems the second approach would give user some more flexibility.

**Comment by psss** [comment]
Quite a useful feature, easy to implement, we just need to agree on the option naming, tentatively proposing for the next sprint.
